# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [1]:
# Imports section
import pandas
import numpy
import sklearn
from sklearn import linear_model
from sklearn import neural_network
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
science_data = pandas.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
print(science_data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(science_data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science_data.drop(columns=["Size nm^3"])  # Assuming 'growth' is the label
y = science_data["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.1, random_state=42, shuffle = True)

## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
lin_reg = sklearn.linear_model.LinearRegression()
lin_reg.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = numpy.array([[404, 309]])
prediction = lin_reg.predict(sample_data)
print("Sample Datapoint: [[404],[309]]. Prediction:", prediction[0])
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print("Linear Regression Score:", lin_reg.score(X_test, y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficients:", lin_reg.coef_)
print("Intercept:", lin_reg.intercept_)

Sample Datapoint: [[404],[309]]. Prediction: 259634.44696295937
Linear Regression Score: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340833


C:\Users\local1\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = −409391.48+866.15⋅x_1+1032.70⋅x_2 $, where $x_1$ is Temperature, $x_2$ is Mols of KCl 

Report on score and explain meaning:

The Linear Regression Score is 0.8552, which represents the $R^2$ score. This value indicates how well the model explains the variance in the target variable.

A score of 1.0 means the model perfectly predicts the output.

A score of 0.8552 suggests that 85.52% of the variation in the slime size can be explained by the temperature and the mols of KCl.

The remaining 14.48% is due to factors not captured by this model, such as noise or missing variables.

Overall, the model performs well, but it may still have room for improvement, potentially through polynomial regression or adding more features.

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
k_fold = KFold(n_splits=5, shuffle=True, random_state=42)  
cv_scores = sklearn.model_selection.cross_val_score(lin_reg, X, y, cv=k_fold, scoring="r2")
# Report on their finding and their significance
print("Cross-validation scores:", cv_scores)
print("Mean CV Score:", numpy.mean(cv_scores))

Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean CV Score: 0.8597294202684646


Write findings here:

The $R^2$ values across all five folds are relatively high, ranging from 0.8392 to 0.8720, which suggests to me  that the model performs well across different subsets of the data. The mean cross-validation score of 0.8568 confirms that the model generalizes well beyond the training data.

The small variation in scores suggests the model is stable, meaning its performance is not significantly affected by different data splits.

Since cross-validation helps mitigate overfitting, this result reassures us that the linear regression model is reliable and generalizable.

The mean CV score of 0.8568 is close to the original test set score of 0.8552, which means our model is not overly dependent on a particular train-test split.

The CV results confirm that the linear regression model is pretty effective. However, the consistent unexplained variance suggests that additional factors could improve the model, such as polynomial terms or additional features.

## 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

poly = sklearn.preprocessing.PolynomialFeatures(degree = 2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

poly_reg = sklearn.linear_model.LinearRegression()
poly_reg.fit(X_train_poly, y_train)

print("Polynomial Regression Score:", poly_reg.score(X_test_poly, y_test))


# Perform k-fold cross validation (as above)
X_poly = poly.fit_transform(X)  
k_fold = KFold(n_splits=5, shuffle=True, random_state=42)  
cv_scores_poly = sklearn.model_selection.cross_val_score(poly_reg, X_poly, y, cv=k_fold, scoring="r2")

# Output cross-validation scores
print("Polynomial Regression Cross-Validation Scores:", cv_scores_poly)
print("Mean CV Score:", numpy.mean(cv_scores_poly))




# Report on the metrics and output the resultant equation as you did in Part 3.

print("Coefficients:", poly_reg.coef_)
print("Intercept:", poly_reg.intercept_)



Polynomial Regression Score: 1.0
Polynomial Regression Cross-Validation Scores: [1. 1. 1. 1. 1.]
Mean CV Score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27197810e-07  1.26444795e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0478852093219757e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$h(x) = 2.0479 \times 10^{-5} + (12.0000 \cdot x_1) + (-1.272 \times 10^{-7} \cdot x_2) + (1.264 \times 10^{-11} \cdot x_1^2) + (2.0000 \cdot x_2^2) + (0.02857 \cdot x_1 \cdot x_2)$, where $x_1$ is Temperature, $x_2$ is Mols of KCl 


Report on the score and interpret:

The polynomial regression model achieved a perfect $R^2$ score of 1.0 both on the test set and in CV. This is concerning to me because it suggests that we are overfitting the data. In comparison, the linear regression model had an $R^2$ score of 0.8552, meaning it explained 85.52% of the variance, while the remaining 14.48% was unexplained.The linear regression model, while slightly less accurate, may be a more generalizable choice for making predictions on unseen data. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [7]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pandas.read_csv("ckd_feature_subset.csv")
X_ckd = ckd_data.drop(columns=["Target_ckd"])  
y_ckd = ckd_data["Target_ckd"]
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X_ckd, y_ckd, test_size=0.1, random_state=42)

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "SVM": SVC(kernel="rbf", random_state=42),
    "k-NN": KNeighborsClassifier(n_neighbors=5),
    "Neural Network": MLPClassifier(hidden_layer_sizes=(50,), max_iter=1000, random_state=42)
}

# Perform 5-fold cross-validation and store results
for name, model in models.items():
    k_fold = KFold(n_splits=5, shuffle=True, random_state=42)  
    cv_scores = sklearn.model_selection.cross_val_score(model, X_ckd, y_ckd, cv=k_fold, scoring="accuracy")
    mean_score = numpy.mean(cv_scores)
    std_dev = numpy.std(cv_scores)
    print(f"{name}: Mean Accuracy = {mean_score}, Std Dev = {std_dev}")

Logistic Regression: Mean Accuracy = 0.8565591397849461, Std Dev = 0.0662686697755876
SVM: Mean Accuracy = 0.9281720430107526, Std Dev = 0.04760126519964333
k-NN: Mean Accuracy = 0.9279569892473118, Std Dev = 0.05243961560473918


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepd

Neural Network: Mean Accuracy = 0.9350537634408603, Std Dev = 0.040465552982107675


## Results and Conclusion for Classification Experiments
| Model                | Mean Accuracy | Std Dev  |
|----------------------|--------------|----------|
| Logistic Regression | 0.8566        | 0.0663   |
| SVM                 | 0.9282        | 0.0476   |
| k-NN                | 0.9280        | 0.0524   |
| Neural Network      | 0.9351        | 0.0405   |

---

The neural network achieved the highest mean accuracy (0.9351) and had the lowest standard deviation (0.0405), indicating strong performance and stability. The SVM model came in second, with an accuracy of 0.9282, performing almost as well but with slightly higher variance.

Logistic regression had the lowest accuracy (0.8566), likely due to its linear decision boundary, which may not capture the complexity of the dataset as well as nonlinear models like SVM or neural networks.

The k-NN model performed similarly to SVM, with an accuracy of 0.9280, but had a higher standard deviation (0.0524), suggesting it was less stable across different splits of the data.

The neural network likely performed best because it can model complex, nonlinear decision boundaries better than logistic regression. The use of hidden layers allowed it to learn intricate patterns in the data that other models could not capture. Its lower standard deviation indicates that its performance was more consistent across different folds of the data.

Neural networks performed the best overall, balancing both high accuracy and low variance. SVM and k-NN were strong contenders, but k-NN had more variability. Logistic regression had the weakest performance, likely due to its linear nature.

In [8]:
nn_configs = [
    ((50,), 0.0001, 0.001, 200, 0.9),  #baseline
    #hidden layers testing
    ((100,), 0.0001, 0.001, 200, 0.9),
    ((25,), 0.0001, 0.001, 200, 0.9),
    ((5,), 0.0001, 0.001, 200, 0.9),
    ((100,100), 0.0001, 0.001, 200, 0.9),
    ((50,50), 0.0001, 0.001, 200, 0.9),
    ((25,25), 0.0001, 0.001, 200, 0.9),
    ((5,5), 0.0001, 0.001, 200, 0.9),
    ((5,50,100), 0.0001, 0.001, 200, 0.9),
    ((100,50,5), 0.0001, 0.001, 200, 0.9),
    #alpha testing
    ((50,), 0.001, 0.001, 200, 0.9),
    ((50,), 0.01, 0.001, 200, 0.9),
    ((50,), 0.1, 0.001, 200, 0.9),
    ((50,), 0.00001, 0.001, 200, 0.9),
    #initial learning rate testing
    ((50,), 0.0001, 0.0001, 200, 0.9),
    ((50,), 0.0001, 0.01, 200, 0.9),
    ((50,), 0.0001, 0.1, 200, 0.9),
    ((50,), 0.0001, 0.00001, 200, 0.9),
    #max iterations testing
    ((50,), 0.0001, 0.001, 500, 0.9),
    ((50,), 0.0001, 0.001, 1000, 0.9),
    ((50,), 0.0001, 0.001, 100, 0.9),
    ((50,), 0.0001, 0.001, 50, 0.9),
    ((50,), 0.0001, 0.001, 1500, 0.9),
    ((50,), 0.0001, 0.001, 2000, 0.9),
    ((50,), 0.0001, 0.001, 5000, 0.9),
    ((50,), 0.0001, 0.001, 10000, 0.9),
    #momentum testing
    ((50,), 0.0001, 0.001, 200, 0.7),
    ((50,), 0.0001, 0.001, 200, 0.5),
    ((50,), 0.0001, 0.001, 200, 0.3),
    ((50,), 0.0001, 0.001, 200, 0.1),
    ((50,), 0.0001, 0.001, 200, 0.01),
]

for config in nn_configs:
    hidden_layers, alpha, lr_init, max_iterations, mom = config
    nn = sklearn.neural_network.MLPClassifier(
        hidden_layer_sizes=hidden_layers,
        alpha=alpha, 
        learning_rate_init=lr_init, 
        max_iter=max_iterations, 
        momentum = mom,
        random_state=42
    )
    
    nn.fit(X_train, y_train)
    y_pred = nn.predict(X_test)
    acc = sklearn.metrics.accuracy_score(y_test, y_pred)
    
    print(f"Neural Network {config} Accuracy: {acc:.4f}")

Neural Network ((50,), 0.0001, 0.001, 200, 0.9) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((100,), 0.0001, 0.001, 200, 0.9) Accuracy: 0.7500
Neural Network ((25,), 0.0001, 0.001, 200, 0.9) Accuracy: 0.6875
Neural Network ((5,), 0.0001, 0.001, 200, 0.9) Accuracy: 0.6250


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((100, 100), 0.0001, 0.001, 200, 0.9) Accuracy: 0.8750


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50, 50), 0.0001, 0.001, 200, 0.9) Accuracy: 0.8750
Neural Network ((25, 25), 0.0001, 0.001, 200, 0.9) Accuracy: 0.7500


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((5, 5), 0.0001, 0.001, 200, 0.9) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((5, 50, 100), 0.0001, 0.001, 200, 0.9) Accuracy: 0.8750


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((100, 50, 5), 0.0001, 0.001, 200, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.001, 0.001, 200, 0.9) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.01, 0.001, 200, 0.9) Accuracy: 0.6875
Neural Network ((50,), 0.1, 0.001, 200, 0.9) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 1e-05, 0.001, 200, 0.9) Accuracy: 0.6875
Neural Network ((50,), 0.0001, 0.0001, 200, 0.9) Accuracy: 0.6250


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.0001, 0.01, 200, 0.9) Accuracy: 0.9375
Neural Network ((50,), 0.0001, 0.1, 200, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 1e-05, 200, 0.9) Accuracy: 0.6250


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.0001, 0.001, 500, 0.9) Accuracy: 0.8125
Neural Network ((50,), 0.0001, 0.001, 1000, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 0.001, 100, 0.9) Accuracy: 0.6875
Neural Network ((50,), 0.0001, 0.001, 50, 0.9) Accuracy: 0.6250


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (50) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.0001, 0.001, 1500, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 0.001, 2000, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 0.001, 5000, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 0.001, 10000, 0.9) Accuracy: 0.8750
Neural Network ((50,), 0.0001, 0.001, 200, 0.7) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.0001, 0.001, 200, 0.5) Accuracy: 0.6875
Neural Network ((50,), 0.0001, 0.001, 200, 0.3) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network ((50,), 0.0001, 0.001, 200, 0.1) Accuracy: 0.6875
Neural Network ((50,), 0.0001, 0.001, 200, 0.01) Accuracy: 0.6875


C:\Users\local1\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

| Hidden Layers       | Alpha  | Learning Rate | Max Iterations | Momentum | Accuracy |
|---------------------|--------|---------------|----------------|----------|----------|
| (50,)              | 0.0001 | 0.001         | 200            | 0.9      | 0.6875   |
| (100,)             | 0.0001 | 0.001         | 200            | 0.9      | 0.7500   |
| (25,)              | 0.0001 | 0.001         | 200            | 0.9      | 0.6875   |
| (5,)               | 0.0001 | 0.001         | 200            | 0.9      | 0.6250   |
| (100, 100)         | 0.0001 | 0.001         | 200            | 0.9      | 0.8750   |
| (50, 50)           | 0.0001 | 0.001         | 200            | 0.9      | 0.8750   |
| (25, 25)           | 0.0001 | 0.001         | 200            | 0.9      | 0.7500   |
| (5, 5)             | 0.0001 | 0.001         | 200            | 0.9      | 0.6875   |
| (5, 50, 100)       | 0.0001 | 0.001         | 200            | 0.9      | 0.8750   |
| (100, 50, 5)       | 0.0001 | 0.001         | 200            | 0.9      | 0.8750   |
| (50,)              | 0.0001 | 0.01          | 200            | 0.9      | 0.9375   |
| (50,)              | 0.0001 | 0.1           | 200            | 0.9      | 0.8750   |
| (50,)              | 0.0001 | 0.0001        | 200            | 0.9      | 0.6250   |
| (50,)              | 0.0001 | 0.001         | 1000           | 0.9      | 0.8750   |
| (50,)              | 0.0001 | 0.001         | 5000           | 0.9      | 0.8750   |
| (50,)              | 0.0001 | 0.001         | 10000          | 0.9      | 0.8750   |
| (50,)              | 0.0001 | 0.001         | 200            | 0.7      | 0.6875   |
| (50,)              | 0.0001 | 0.001         | 200            | 0.3      | 0.6875   |
| (50,)              | 0.0001 | 0.001         | 200            | 0.01     | 0.6875   |

---

Increasing the number of layers and neurons improved performance. Single-layer networks with fewer neurons, like (5,) or (25,) performed worse, with accuracies around 0.6250-0.6875. Networks with two or three layers, like (100,100), (50,50), or (5,50,100) achieved higher accuracy, around 0.8750.

The learning rate had a big impact on performance. Higher learning rates, such as 0.01, performed best, achieving an accuracy of 0.9375. Extremely small or large learning rates, such as 0.0001 or 0.1, caused performance drops.

More training iterations helped convergence. With 200 iterations, many tests showed "Maximum iterations reached" warnings. Increasing to 1000+ iterations improved accuracy, from 0.6875 to 0.8750.

Momentum had little impact on accuracy. Changing momentum values from 0.9 to 0.7, 0.3, and 0.01 did not affect accuracy much, which stayed around 0.6875.

In conclusion, the best configuration was (50,), with a 0.0001 alpha, 0.01 learning rate, 200 iterations, and 0.9 momentum, achieving 93.75% accuracy. The biggest impact came from the learning rate and the number of hidden layers. A potential future improvement would be to increase max_iter further, to 2000 or more, to see if accuracy continues improving.